### PR TITLE
Gobierto Visualizations / Subsidies issues

### PR DESCRIPTION
**Pull request by jorgeatgu**
Closes https://github.com/PopulateTools/issues/issues/1149

## :v: What does this PR do?

- Create a unique id for each beneficiary
- Filter counters are updated at every interaction.
- Show the full name of each beneficiary.
- Show zero instead of `NaN`

## :mag: How should this be manually tested?

Staging

### Before this PR

before

### After this PR

after

### CODE REVIEW

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js [hunk 1]
```diff
@@ -118,6 +118,12 @@ export class SubsidiesController {
   setGlobalVariables(rawData) {
     let subsidiesData = rawData[0];
 
+    subsidiesData.sort((a, b) => {
+      a = a.grant_date.replace(/-/g, '');
+      b = b.grant_date.replace(/-/g, '');
+      return a.localeCompare(b)
+    });
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: What are you doing with the replace clauses exactly? Have you tried if any localCompare method _options_ works to you? Check the [available options](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Collator/Collator).

On the other hand, what if is there no `grant_date`? 💥 what if `grant_date` is not an string? 💥

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js [hunk 2]
```diff
@@ -135,23 +141,17 @@ export class SubsidiesController {
           ? parseFloat(subsidy.amount)
           : 0.0;
 
-      let [beneficiary_id, ...beneficiary_name] = subsidy.beneficiary.split(
-        " "
-      );
+      let [beneficiary_id, ...beneficiary_name] = subsidy.beneficiary.split(" ");
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Are you sure `beneficiary` always come as string?

Actually this whole `for` block, deserves an important refactorization, replacing the `for` loop with a `map`, and providing fallbacks for every single prop. For instance:

```js
subsidiesData.map(({ amount = 0, beneficiary = "", ...rest }) => {
  // any complex logic you need before returning

  return {
    ...rest,
    amount: ...,
    range: ...,
    // etc...
  }
})
```

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js [hunk 3]
```diff
@@ -203,32 +203,32 @@ export class SubsidiesController {
 
     // Calculations box items
     const numberSubsidies = subsidiesData.length;
-    const sumSubsidies = d3.sum(amountsArray);
+    const sumSubsidies = d3.sum(amountsArray) || 0;
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: What if you create a helper function to return all this calculations? Here's an obvious DRY

File: app/javascript/gobierto_visualizations/webapp/containers/subsidies/SubsidiesIndex.vue [hunk 4]
```diff
@@ -47,9 +47,9 @@ export default {
     updateFilteredItems(value) {
       this.value = value
       if (!this.value) {
-        this.items = this.subsidiesData.filter(contract => contract.beneficiary_name.toLowerCase().includes(this.value.toLowerCase()))
+        this.items = this.subsidiesData.filter(contract => contract.beneficiary.toLowerCase().includes(this.value.toLowerCase()))
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/webapp/containers/subsidies/SubsidiesIndex.vue]: Always that you apply functions/mutations to something you've to provide fallback, in this case destructuring.

```suggestion
        this.items = this.subsidiesData.filter(({ beneficiary = "" }) => beneficiary.toLowerCase().includes(this.value.toLowerCase()))
```

File: app/javascript/gobierto_visualizations/webapp/containers/subsidies/SubsidiesIndex.vue [hunk 5]
```diff
@@ -47,9 +47,9 @@ export default {
...
       } else {
-        this.items = this.subsidiesData.filter(contract => contract.beneficiary_name.toLowerCase().includes(this.value.toLowerCase())).slice(0, 25)
+        this.items = this.subsidiesData.filter(contract => contract.beneficiary.toLowerCase().includes(this.value.toLowerCase())).slice(0, 25)
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/webapp/containers/subsidies/SubsidiesIndex.vue]: Same

**Review by Crashillo — CHANGES_REQUESTED** [review]

**Comment by ferblape** [comment]
For subsidies with blank beneficiary and with tipo_beneficiario == 'persona', could you use the string "Persona individual" both in the list and the show action? It's weird to have blanks, we should explain why are they blank

Another question, why have you merged improvement_visualizations branch here?

**Comment by jorgeatgu** [comment]
> For subsidies with blank beneficiary and with tipo_beneficiario == 'persona', could you use the string "Persona individual" both in the list and the show action? It's weird to have blanks, we should explain why are they blank

Done

> Another question, why have you merged improvement_visualizations branch here?

To prevent conflicts

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js (hunk 1 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Done!

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js (hunk 2 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Done!

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js (hunk 3 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Done!

File: app/javascript/gobierto_visualizations/webapp/containers/subsidies/SubsidiesIndex.vue (hunk 4 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/webapp/containers/subsidies/SubsidiesIndex.vue]: Done!

File: app/javascript/gobierto_visualizations/webapp/containers/subsidies/SubsidiesIndex.vue (hunk 5 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/webapp/containers/subsidies/SubsidiesIndex.vue]: Done!

File: app/javascript/gobierto_visualizations/modules/contracts_controller.js [hunk 6]
```diff
@@ -256,9 +263,9 @@ export class ContractsController {
     );
 
     const numberTenders = _tendersData.length;
-    const sumTenders = d3.sum(amountsArray);
-    const meanTenders = d3.mean(amountsArray);
-    const medianTenders = d3.median(amountsArray);
+    const sumTenders = d3.sum(amountsArray) || 0;
+    const meanTenders = d3.mean(amountsArray) || 0;
+    const medianTenders = d3.median(amountsArray) || 0;
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/modules/contracts_controller.js]: May it sense use here also `calculateSumMeanMedian`? As well as contracts

File: app/javascript/gobierto_visualizations/modules/contracts_controller.js [hunk 7]
```diff
@@ -295,7 +302,7 @@ export class ContractsController {
       _contractsData,
       ({ final_amount_no_taxes = 0 }) => parseFloat(final_amount_no_taxes)
     );
-    const largerContractAmountPct = largerContractAmount / sumContracts;
+    const largerContractAmountPct = (largerContractAmount / sumContracts) || 0;
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/modules/contracts_controller.js]: Possible division by zero

```suggestion
    const largerContractAmountPct = sumContracts ? (largerContractAmount / sumContracts) : 0;
```

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js [hunk 8]
```diff
@@ -128,32 +128,25 @@ export class SubsidiesController {
...
-    }
+
+      return {
+        amount: amount && !Number.isNaN(amount) ? parseFloat(amount) : 0.0,
+        range: rangeFormat(+amount),
+        beneficiary: beneficiary,
+        beneficiary_type: beneficiary_type,
+        beneficiary_id: beneficiary_id,
+        beneficiary_name: beneficiary_name,
+        grant_date: grant_date,
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Easier:

```suggestion
        beneficiary,
        beneficiary_type,
        beneficiary_id,
        beneficiary_name,
        grant_date,
```

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js [hunk 9]
```diff
@@ -203,32 +196,26 @@ export class SubsidiesController {
...
-    const meanSubsidies = d3.mean(amountsArray) || 0;
-    const medianSubsidies = d3.median(amountsArray) || 0;
+    const [ sumSubsidies, meanSubsidies, medianSubsidies ] = calculateSumMeanMedian(amountsArray)
 
     const pctCollectivesSubsidies =
-      parseFloat(collectivesData.length) / numberSubsidies;
-    const sumCollectivesSubsidies = d3.sum(amountsCollectivesArray);
-    const meanCollectivesSubsidies = d3.mean(amountsCollectivesArray) || 0;
-    const medianCollectivesSubsidies = d3.median(amountsCollectivesArray) || 0;
+      (parseFloat(collectivesData.length) / numberSubsidies) || 0;
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Division by zero

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js [hunk 10]
```diff
@@ -203,32 +196,26 @@ export class SubsidiesController {
...
+    const [ sumCollectivesSubsidies, meanCollectivesSubsidies, medianCollectivesSubsidies ] = calculateSumMeanMedian(amountsCollectivesArray)
 
     const pctIndividualsSubsidies =
-      parseFloat(individualsData.length) / numberSubsidies;
-    const sumIndividualsSubsidies = d3.sum(amountsIndividualsArray);
-    const meanIndividualsSubsidies = d3.mean(amountsIndividualsArray) || 0;
-    const medianIndividualsSubsidies = d3.median(amountsIndividualsArray) || 0;
+      (parseFloat(individualsData.length) / numberSubsidies) || 0;
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Division by zero

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js [hunk 11]
```diff
@@ -203,32 +196,26 @@ export class SubsidiesController {
...
+    const [ sumIndividualsSubsidies, meanIndividualsSubsidies, medianIndividualsSubsidies ] = calculateSumMeanMedian(amountsIndividualsArray)
 
     // Calculations headlines
     const lessThan1000Total = subsidiesData.filter(
       ({ amount = 0 }) => parseFloat(amount) < 1000
     ).length;
-    const lessThan1000Pct = lessThan1000Total / numberSubsidies;
+    const lessThan1000Pct = (lessThan1000Total / numberSubsidies) || 0;
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Division by zero

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js [hunk 12]
```diff
@@ -203,32 +196,26 @@ export class SubsidiesController {
...
 
     const largerSubsidyAmount = d3.max(subsidiesData, ({ amount = 0 }) =>
       parseFloat(amount)
     );
-    const largerSubsidyAmountPct = largerSubsidyAmount / sumSubsidies;
+    const largerSubsidyAmountPct = (largerSubsidyAmount / sumSubsidies) || 0;
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Division by zero

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js [hunk 13]
```diff
@@ -240,8 +227,7 @@ export class SubsidiesController {
         break;
       }
     }
-    const halfSpendingsSubsidiesPct =
-      numberSubsidiesHalfSpendings / numberSubsidies;
+    const halfSpendingsSubsidiesPct = (numberSubsidiesHalfSpendings / numberSubsidies) || 0;
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Division by zero

File: app/javascript/lib/vue-components/modules/BeesWarmChart.vue [hunk 14]
```diff
@@ -100,13 +105,25 @@ export default {
         this.deepCloneData(newValue);
         this.updateData = true;
       }
+    },
+    $route(to, from) {
+      if (to !== from) {
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/BeesWarmChart.vue]: This comparison is worthless. Objects cannot be compared like primitives

File: app/javascript/lib/vue-components/modules/BeesWarmChart.vue [hunk 15]
```diff
@@ -270,9 +280,12 @@ export default {
           d.id = normalizeString(d.id)
           return `beeswarm-circle beeswarm-circle-${d.id} beeswarm-circle-${d.slug_contract_type}`
         })
+        .attr('id', d => `${d.slug}`)
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/BeesWarmChart.vue]: Why not just?
```suggestion
        .attr('id', d => d.slug)
```

File: app/javascript/gobierto_visualizations/webapp/visualizations/treeMapNested.vue [hunk 16]
```diff
@@ -172,10 +178,11 @@ export default {
       d3.select('.treemap-container')
         .remove()
         .exit();
-      d3.select('.nav-breadcrumbs')
+      d3.select('.treemap-nested-sidebar-nav-breadcumb')
         .remove()
         .exit();
 
+      const colors = createScaleColors(this.arrayValuesContractTypes.length, this.arrayValuesContractTypes);
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/webapp/visualizations/treeMapNested.vue]: What if `this.arrayValuesContractTypes.length` is longer than the `colorsGobiertoExtend` palette in _createScaleColors_?

File: app/javascript/gobierto_visualizations/webapp/visualizations/treeMapNested.vue [hunk 17]
```diff
@@ -229,36 +237,19 @@ export default {
...
-            let parentNameContract = parentName
-            let sectionName = name
-            //Normalize and create an slug because Servicios de Gestión Públicos isn't a valid css class
-            sectionName = normalizeString(sectionName)
-            parentNameContract = normalizeString(parentNameContract)
-            return `children ${sectionName} ${parentNameContract}`
+          .attr('class', 'children')
+          .attr('fill', d => {
+            const { depth, data: { name = '' }, parent: { data: { name: parentName = '' } } } = d
+            const valueColor = depth === 1 ? d.color = colors(name) : d.color = colors(parentName)
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/webapp/visualizations/treeMapNested.vue]: Mmmm, you're setting `d.color`, but you return `valueColor`. Is that any sense?

File: app/javascript/gobierto_visualizations/webapp/visualizations/treeMapNested.vue [hunk 18]
```diff
@@ -377,39 +369,40 @@ export default {
           })
 
         g.selectAll(".foreignobj")
-        .on('mousemove', function(d) {
+        .on('mousemove', function(d, i) {
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/webapp/visualizations/treeMapNested.vue]: What for `i`?

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/lib/vue-components/modules/BeesWarmChart.vue [hunk 19]
```diff
@@ -14,7 +18,7 @@ import { timeParse, timeFormat, timeFormatLocale } from 'd3-time-format';
 import { min, max } from 'd3-array';
 import { d3locale } from 'lib/shared';
 import { easeLinear } from 'd3-ease';
-import { normalizeString } from "../../../gobierto_visualizations/webapp/lib/utils";
+import { normalizeString, createScaleColors } from "../../../gobierto_visualizations/webapp/lib/utils";
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/BeesWarmChart.vue]: Wait, I forgot this. Totally wrong. You're setting a dependency between a component and a module

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/lib/vue-components/modules/BeesWarmChart.vue (hunk 19 above)
Comment (Crashillo) [reply app/javascript/lib/vue-components/modules/BeesWarmChart.vue]: Everything declared in _lib/shared_ (or lib/whatever) must work ON HIS OWN.

Comment (jorgeatgu) [reply app/javascript/lib/vue-components/modules/BeesWarmChart.vue]: Done!

File: app/javascript/gobierto_visualizations/webapp/visualizations/treeMapNested.vue (hunk 18 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/webapp/visualizations/treeMapNested.vue]: Done!

File: app/javascript/gobierto_visualizations/webapp/visualizations/treeMapNested.vue (hunk 17 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/webapp/visualizations/treeMapNested.vue]: Done!

File: app/javascript/lib/vue-components/modules/BeesWarmChart.vue (hunk 15 above)
Comment (jorgeatgu) [reply app/javascript/lib/vue-components/modules/BeesWarmChart.vue]: Done!

File: app/javascript/lib/vue-components/modules/BeesWarmChart.vue (hunk 14 above)
Comment (jorgeatgu) [reply app/javascript/lib/vue-components/modules/BeesWarmChart.vue]: Done!

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js (hunk 13 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Done!

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js (hunk 12 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Done!

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js (hunk 11 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Done!

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js (hunk 10 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Done!

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js (hunk 9 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Done!

File: app/javascript/gobierto_visualizations/modules/subsidies_controller.js (hunk 8 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/modules/subsidies_controller.js]: Done!

File: app/javascript/gobierto_visualizations/modules/contracts_controller.js (hunk 7 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/modules/contracts_controller.js]: Done!

File: app/javascript/gobierto_visualizations/modules/contracts_controller.js (hunk 6 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_visualizations/modules/contracts_controller.js]: Done!

File: app/javascript/gobierto_visualizations/webapp/visualizations/treeMapNested.vue (hunk 17 above)
Comment (Crashillo) [reply app/javascript/gobierto_visualizations/webapp/visualizations/treeMapNested.vue]: Wait there. Why are you doing an assignment `d.color = colors(name)` or `d.color = colors(parentName)` inside the returning value (or the previous `valueColor`)? That's what I wanted to pointed out.

File: app/javascript/lib/vue-components/modules/BeesWarmChart.vue (hunk 19 above)
Comment (Crashillo) [reply app/javascript/lib/vue-components/modules/BeesWarmChart.vue]: No man, don't you realise you're moving just `createScaleColors`? What happens with `normalizeString`? It's the same.

You cannot create a dependency between modules (that's [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming))), the wrong thing here is about the path: **../../../gobierto_visualizations/webapp/lib/utils**. Think about it. You're in "lib/vue-components". And you're setting some stuff who comes from _gobierto_visualizations_. Tomorrow I wanna create a BeesWarmChart in _gobierto_budgets_.... and it fetches code from an unrelated module.

**Review by Crashillo — CHANGES_REQUESTED** [review]
I've updated the unresolved comments.

**Comment by Crashillo** [comment]
Merged master and solved conflict with vue-directives

**Review by ferblape — APPROVED** [review]

**Pull request merged**